music_assistant/providers/ard_audiothek/__init__.py

```diff
@@ -0,0 +1,913 @@
+        ),
+    )
+
+
+class ARDAudiothek(MusicProvider):
+    """
+    Example/demo Music provider.
+
+    Note that this is always subclassed from MusicProvider,
+    which in turn is a subclass of the generic Provider model.
+
+    The base implementation already takes care of some convenience methods,
+    such as the mass object and the logger. Take a look at the base class
+    for more information on what is available.
+
+    Just like with any other subclass, make sure that if you override
+    any of the default methods (such as __init__), you call the super() method.
+    In most cases its not needed to override any of the builtin methods and you only
+    implement the abc methods with your actual implementation.
+    """
+
+    @property
+    def supported_features(self) -> set[ProviderFeature]:
+        """Return the features supported by this Provider."""
+        # MANDATORY
+        # you should return a tuple of provider-level features
+        # here that your player provider supports or an empty tuple if none.
+        # for example 'ProviderFeature.SYNC_PLAYERS' if you can sync players.
+        return {
+            ProviderFeature.BROWSE,
+            ProviderFeature.SEARCH,
+            ProviderFeature.RECOMMENDATIONS,
+            ProviderFeature.LIBRARY_RADIOS,
+            ProviderFeature.LIBRARY_PODCASTS,
+            ProviderFeature.LIBRARY_PODCASTS_EDIT,
+            # ProviderFeature.SIMILAR_TRACKS,
+            # see the ProviderFeature enum for all available features
+        }
+
+    async def handle_async_init(self) -> None:
+        """Pass config values to client and initialize."""
+        _email = self.config.get_value(CONF_EMAIL)
+        _password = self.config.get_value(CONF_PASSWORD)
+        self.token = self.config.get_value(CONF_TOKEN_BEARER)
+        self.user_id = self.config.get_value(CONF_USERID)
+        self.token_expire = datetime.fromtimestamp(
+            float(str(self.config.get_value(CONF_EXPIRY_TIME)))
+        )
+
+        self.max_bitrate = int(float(str(self.config.get_value(CONF_MAX_BITRATE))))
+
+        if (
+            _email is not None
+            and _password is not None
+            and (self.token is None or self.user_id is None or self.token_expire < datetime.now())
+        ):
+            self.token, self.user_id, _username = await _login(
+                self.mass.http_session, str(_email), str(_password)
+            )
+            self.update_config_value(CONF_TOKEN_BEARER, self.token, encrypted=True)
+            self.update_config_value(CONF_USERID, self.user_id, encrypted=True)
+            self.update_config_value(CONF_USER_NAME, _username)
+            self.update_config_value(
+                CONF_EXPIRY_TIME, str((datetime.now() + timedelta(hours=1)).timestamp())
+            )
+
+        headers = None
+        if self.token:
+            headers = {"Authorization": f"Bearer {self.token}"}
+
+        self._client = Client(
```

Comment:
Saving the client (including the bearer token) in the `init` will eventually result in the token getting expired. The user will need to restart the provider to solve this. I would suggest to moving this code into a separate function so you can also call it whenever a bearer token has expired. You will need to check this with every api call.